Add more initialization options for the namelist jphgr_msh parameter and fix C++OpenMP flather_v loop

sergisiso opened the pull request:
Issue #48 

This PR adds the possibility to initialize grids with different open boundaries and fixes the discrepancy between OpenMP Fortran and C++ noted by @AidanChalk .

sergisiso commented:
@arporter This is ready for review.

arporter commented:
All looks good although I think there may be an issue with the T mask in the NE corner:

<img width="320" alt="tmask_se" src="https://user-images.githubusercontent.com/17409212/87302637-ae6d5600-c509-11ea-9613-edf6ec647639.png">

sergisiso commented on benchmarks/nemo/nemolite2d/common/model_mod.f90:
```diff
@@ -166,8 +169,14 @@ subroutine setup_tpoints_mask(subdomain, jpi, jpj, tmask)
-          tmask(subdomain%internal%xstop+1:, :) = 0
+          ! Eastern global boundary may be open (tidally forced)
+          ! Make the open boundary outside the computational domain
+          ! (i.e. ystart - 1)
```
You are right, fixed.

arporter replied:
I've taken the liberty of s/xstart/xstop/ .

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step_omp.cpp:
```diff
@@ -115,7 +115,8 @@ extern "C" void c_invoke_time_step(
 
     // Boundary conditions bc_flather_v kernel (whole domain but top y boundary)
-    #pragma omp parallel for
+    // We cannot execute this loop in (OpenMP) parallel because of the
```
I think it is mainly to have a fair comparison with the Fortran OpenMP, but it should have a small impact as flather takes a small percentage of execution time.

Aha, the psyclone version does omp-parallelise this loop, so this will need a 'kernel fix' as you mention at some point.

sergisiso commented:
@arporter I fixed the comment and changed the order of the if conditions in the initialization so the NE corner should be ok now, although it is probably not a problem as you say.

Your are right that the flather_v kernel (and I think the same applies to flather_u) have a definition problem as they have possible dependencies.

Ready for next review.

arporter commented:
The new version builds and runs fine. I've taken the liberty of correcting a remaining typo in a comment. Once Travis is happy I shall merge.
